feat: dfx deps subcommands (#3095)
dfinity/sdk: docs/concepts/pull-dependencies.md

# Pull Dependencies

## Overview

The interoperability of canisters on the Internet Computer (IC) is an important feature. 

> `dfx` provides a consistent developer workflow for integrating thrid party canisters.

Reviewer: ```suggestion
`dfx` provides a consistent developer workflow for integrating third-party canisters.
```

A service provider attaches [necessary metadata](canister-metadata.md) to the canister wasm.

A service consumer then can pull dependencies directly from the IC mainnet and easily deploy them on a local replica.

> This document describes the consumer workflow and explains what happens behind the scene.

Reviewer: ```suggestion
This document describes the consumer workflow and explains what happens behind the scenes.
```

## Workflow

### 1. Declare "pull" dependencies in `dfx.json`

Below is an example `dfx.json`, the service consumer is developing the "app" canister which has two pull dependencies:

- "dep_b" which has canister ID of "yhgn4-myaaa-aaaaa-aabta-cai" on the mainnet.
- "dep_c" which has canister ID of "yahli-baaaa-aaaaa-aabtq-cai" on the mainnet.

```json
{
    "canisters": {
        "app": {
            "type": "motoko",
            "main": "src/main.mo",
            "dependencies": [
                "dep_b", "dep_c"
            ]
        },
        "dep_b": {
            "type": "pull",
            "id": "yhgn4-myaaa-aaaaa-aabta-cai"
        },
        "dep_c": {
            "type": "pull",
            "id": "yahli-baaaa-aaaaa-aabtq-cai"
        }
    }
}
```

### 2. Pull the dependencies using `dfx deps pull`

> Running `dfx deps pull --network ic` will:

Reviewer: Shouldn't `--network ic` be the default for `dfx deps pull`?  This would be the normal use case.  The other use cases would be `--network local` for CI, and `--network <some test net>` for test networks.

```suggestion
Running `dfx deps pull` will:
```

> 1. resolve the dependency graph by fetching `dfx:deps` metadata recuresively;

Reviewer: ```suggestion
1. resolve the dependency graph by fetching `dfx:deps` metadata recursively;
```

2. download wasm of all direct and indirect dependencies from `dfx:wasm_url` into shared cache;

> 3. hash check the downloaded wasm against `dfx:wasm_hash` metadata or the hash of the mainnet running canister;

Reviewer: ```suggestion
3. verify the hash of the downloaded wasm against `dfx:wasm_hash` metadata or the hash of the canister deployed on mainnet;
```

4. extract `candid:args`, `candid:service`, `dfx:init` from the downloaded wasm;
5. create `deps/` folder in project root;
6. save `candid:service` of direct dependencies as `deps/<NAME>.did`;
7. save `deps/pulled.json` which contains major info of all direct and indirect dependencies;

For the example project, you will find following files in `deps/`:

- `dep_b.did` and `dep_c.did`: candid files that can be imported by "app";
- `pulled.json` which has follwing content:

```
{
  "canisters": {
    "yofga-2qaaa-aaaaa-aabsq-cai": {

>       "name": null,

Reviewer: Please make it so the serializer omits this field if None, rather than writing it as `null` in the JSON.

      "deps": [],
      "wasm_url": "example.com/a.wasm",
      "wasm_hash": "e9b8ba2ad28fa1403cf6e776db531cdd6009a8e5cac2b1097d09bfc65163d56f",
      "dfx_init": "Nat",
      "candid_args": "(nat)"
    },
    "yhgn4-myaaa-aaaaa-aabta-cai": {
      "name": "dep_b",
      "deps": [
        "yofga-2qaaa-aaaaa-aabsq-cai"
      ],
      "wasm_url": "example.com/b.wasm",
      "wasm_hash": "f607c30727b0ee81317fc4547a8da3cda9bb9621f5d0740806ef973af5b479a2",
      "dfx_init": null,
      "candid_args": "()"
    },
    "yahli-baaaa-aaaaa-aabtq-cai": {
      "name": "dep_c",
      "deps": [
        "yofga-2qaaa-aaaaa-aabsq-cai"
      ],
      "wasm_url": "example.com/c.wasm",
      "wasm_hash": "016df9800dc5760785646373bcb6e6bb530fc17f844600991a098ef4d486cf0b",
      "dfx_init": "Nat",
      "candid_args": "(nat)"
    }
  }
}
```

There are three dependencies:

- "yhgn4-myaaa-aaaaa-aabta-cai": "dep_b" in `dfx.json`;
- "yahli-baaaa-aaaaa-aabtq-cai": "dep_c" in `dfx.json`;
- "yofga-2qaaa-aaaaa-aabsq-cai": an indirect dependency that both "dep_b" and "dep_c" depend on;

### 3. Set init arguments using `dfx deps init`

> Running `dfx deps init` will iterate all dependencies in `pulled.json`, set empty argument for the ones need no init argument and print the list of dependencies that require init argument.

Reviewer: ```suggestion
Running `dfx deps init` will iterate over all dependencies in `pulled.json`, set an empty argument for the ones that need no init argument and print the list of dependencies that do require an init argument.
```

Then running `dfx deps init <CANISTER> --argument <ARGUMENT>` to set init argument for individual dependency.

The init arguments will be recorded in `deps/init.json`.

For our example, we should run:

```
> dfx deps init
WARN: The following canister(s) require an init argument. Please run `dfx deps init <NAME/PRINCIPAL>` to set them individually:

> yofga-2qaaa-aaaaa-aabsq-cai

Reviewer: Note that the `dfx:init` metadata in the canister that referenced this indirect dependency does have a name for this dependency, but we discarded it.  Can display it here in an unambiguous way?

Author: In my e2e example project, both `b` and `c` depends on `a`. It's possible that `b` and `c` set different names for `a` in their `dfx:deps`. Which name should we use here?

Another idea is to remove the `name` from `dfx:deps` so that it only contains a list of principals.

Reviewer: I wondered what to do about the potentially different names too, but I don't have a clear answer either.

yahli-baaaa-aaaaa-aabtq-cai (dep_c)

> > dfx deps init yofga-2qaaa-aaaaa-aabsq-cai --argument 10

Reviewer: It feels like we're skipping a step here: how does the developer know what to pass for the `--argument`? 

Would they first run `dfx deps init yofga-2qaaa-aaaaa-aabsq-cai` (not specifying an argument) which would display the `dfx:init` message and/or the `candid:args` ?

    "yhgn4-myaaa-aaaaa-aabta-cai": {
      "arg_str": null,
      "arg_raw": null
    },
    "yahli-baaaa-aaaaa-aabtq-cai": {
      "arg_str": "20",
      "arg_raw": "4449444c00017d14"
    }
  }
}
```

### 4. Deploy pull dependencies on local replica using `dfx deps deploy`

Running `dfx deps deploy` will:

1. create the dependencies on the local replica with the same mainnet canister ID;
2. install the downloaded wasm with the init arguments in `init.json`;

You can also specify the name or principal to deploy one particular dependency.

For our example:
```
> dfx deps deploy
Creating canister: yofga-2qaaa-aaaaa-aabsq-cai
Installing canister: yofga-2qaaa-aaaaa-aabsq-cai
Creating canister: yhgn4-myaaa-aaaaa-aabta-cai (dep_b)
Installing canister: yhgn4-myaaa-aaaaa-aabta-cai (dep_b)
Creating canister: yahli-baaaa-aaaaa-aabtq-cai (dep_c)
Installing canister: yahli-baaaa-aaaaa-aabtq-cai (dep_c)
> dfx deps deploy yofga-2qaaa-aaaaa-aabsq-cai
Creating canister: yofga-2qaaa-aaaaa-aabsq-cai
Installing canister: yofga-2qaaa-aaaaa-aabsq-cai
> dfx deps deploy dep_b
Creating canister: yhgn4-myaaa-aaaaa-aabta-cai (dep_b)
Installing canister: yhgn4-myaaa-aaaaa-aabta-cai (dep_b)
```

**Note**

> - `dfx deps deploy` always create the canister with the anonymous identity so that dependencies and application canisters will have different controllers;

Reviewer: ```suggestion
- `dfx deps deploy` always creates the canister with the anonymous identity so that dependencies and application canisters will have different controllers;
```

> - `dfx deps deploy` always install the canister in "reinstall" mode so that the canister status will be discarded;

Reviewer: ```suggestion
- `dfx deps deploy` always installs the canister in "reinstall" mode so that the canister status will be discarded;
```

A service consumer then can pull dependencies directly from mainnet and easily deploy them on a local replica.

This document describes the workflow and explains what happens behind the scenes.

## Service Provider Workflow

Below is an example provider `dfx.json` which has a `pullable` "service" canister:

```json
{
    "canisters": {
        "service": {
            "type": "motoko",
            "main": "src/main.mo",
            "pullable": {
                "wasm_url": "http://example.com/a.wasm",
                "wasm_hash": "d180f1e232bafcee7d4879d8a2260ee7bcf9a20c241468d0e9cf4aa15ef8f312",
                "dependencies": [
                    "yofga-2qaaa-aaaaa-aabsq-cai"
                ],
                "init_guide": "A natural number, e.g. 10."
            }
        }
    }
}
```

The `pullable` object will be serialized as a part of the `dfx` metadata and attached to the wasm.

Let's go through the properties of the `pullable` object.

### `wasm_url`

A URL to download canister wasm module which will be deployed locally.

### `wasm_hash` 

SHA256 hash of the wasm module located at `wasm_url`.

> This section is optional.

Reviewer: ```suggestion
This field is optional.
```

        "yofga-2qaaa-aaaaa-aabsq-cai"
      ],
      "wasm_hash": "f607c30727b0ee81317fc4547a8da3cda9bb9621f5d0740806ef973af5b479a2",
      "init_guide": "No init arguments required",
      "candid_args": "()"
    },
    "yahli-baaaa-aaaaa-aabtq-cai": {
      "name": "dep_c",
      "dependencies": [
        "yofga-2qaaa-aaaaa-aabsq-cai"
      ],
      "wasm_hash": "016df9800dc5760785646373bcb6e6bb530fc17f844600991a098ef4d486cf0b",
      "init_guide": "A natural number, e.g. 20.",
      "candid_args": "(nat)"
    }
  }
}
```

There are three dependencies:

- "yhgn4-myaaa-aaaaa-aabta-cai": "dep_b" in `dfx.json`;
- "yahli-baaaa-aaaaa-aabtq-cai": "dep_c" in `dfx.json`;
- "yofga-2qaaa-aaaaa-aabsq-cai": an indirect dependency that both "dep_b" and "dep_c" depend on;

**Note**

-  `dfx deps pull` connects to the IC mainnet by default (`--network ic`).
You can choose other network as usual, e.g. `--network local`.

### 3. Set init arguments using `dfx deps init`

Running `dfx deps init` will iterate over all dependencies in `pulled.json`, set an empty argument for the ones that need no init argument and print the list of dependencies that do require an init argument.

> Then running `dfx deps init <CANISTER> --argument <ARGUMENT>` to set init argument for individual dependency.

Reviewer: ```suggestion
Then running `dfx deps init <CANISTER> --argument <ARGUMENT>` will set the init argument for an individual dependency.
```
